src/controllers/Social.ts

```diff
@@ -361,56 +353,62 @@ export default class Social {
       _nonce = accessKeyView.nonce + BigInt(1); // increment nonce as this will be a new transaction for the access key
     }
 
-    // for each account, check if there is storage, if there isn't, add an action to deposit the minimum storage
-    for (let i = 0; i < uniqueAccountIds.length; i++) {
+    // for each key, check if the signer has been granted write permission for the key
+    for (let i = 0; i < keys.length; i++) {
       if (
-        !(await this._storageBalanceOf({
-          accountId: uniqueAccountIds[i],
+        (keys[i].split('/')[0] || '') !== signer.accountId &&
+        !(await this.isWritePermissionGranted({
+          granteePublicKey: publicKey,
+          key: keys[i],
           signer,
         }))
       ) {
-        actions.push(
-          transactions.functionCall(
-            ChangeMethodEnum.StorageDeposit,
-            {
-              account_id: uniqueAccountIds[i],
-              registration_only: true,
-            } as ISocialDBContractStorageDepositArgs,
-            BigInt(GAS_FEE_IN_ATOMIC_UNITS),
-            BigInt(
-              new BigNumber(MINIMUM_STORAGE_IN_BYTES)
-                .multipliedBy(
-                  new BigNumber(STORAGE_COST_PER_BYTES_IN_ATOMIC_UNITS)
-                )
-                .toFixed()
-            ) // minimum storage cost
-          )
+        throw new KeyNotAllowedError(
+          keys[i],
+          `the supplied public key has not been granted write permission for "${keys[i]}"`
         );
       }
     }
 
-    actions.push(
-      transactions.functionCall(
-        ChangeMethodEnum.Set,
-        {
-          data,
-          ...(refundUnusedDeposit && {
-            options: {
-              refund_unused_deposit: refundUnusedDeposit,
-            },
-          }),
-        } as ISocialDBContractSetArgs,
-        BigInt(GAS_FEE_IN_ATOMIC_UNITS),
-        actions.length <= 0 ? BigInt('1') : BigInt('0')
+    // if the signer is updating their own data, check for minimum storage deposit
+    if (
+      this._uniqueAccountIdsFromKeys(keys).find(
+        (value) => value === signer.accountId
       )
-    );
+    ) {
+      storageBalance = await this._storageBalanceOf({
+        accountId: signer.accountId,
+        signer,
+      });
+
+      // if there is no storage balance, use the minimum storage cost as the deposit
```

Comment:
I think using this logic for deposit. The function call will fail. 

If an account already has enough storage for this new call, it will go smoothly. 
However if not, it will fail. 


We should find out deposit required for the new data and subtract it from user's current balance and use that as deposit. 

See how near-social/VM handles it - https://github.com/NearSocial/VM/blob/6047c6a9b96f3de14e600c1d2b96c432bbb76dd4/src/lib/components/Commit.js#L35

Reply:
What do you mean? This logic is to protect against if there is no storage balance (indicated by the return from `storageBalanceOf` being `null`) so this just adds the minimum storage deposit. Otherwise, the call will fail.

This logic only happens when the signer is the one setting their own data.

Reply:
Imagine a scenario where an account currently has 0.5N available storage on-chain. However the new data that you are inputting would need 1N of storage cost. 

In this case, set call should attach 0.5 deposit. 

According to the code above, if there user has even the tiniest bit available storage, there would be no storage attached, no matter how big the data being set is.